File: devops-agent-iac/anexos/cdk-typescript/lib/constructs/secure-bucket.ts
import * as cdk from 'aws-cdk-lib';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as kms from 'aws-cdk-lib/aws-kms';
import { Construct } from 'constructs';

export interface SecureBucketProps {
  readonly bucketName: string;
  readonly encryptionKey: kms.IKey;
  readonly retain: boolean;
  readonly lifecycleRules?: s3.LifecycleRule[];
}

/**
 * Construct L3 reutilizável: bucket com criptografia KMS, bucket key (economia
 * de chamadas ao KMS), versionamento, bloqueio público total, TLS obrigatório
 * e ownership enforced. Padroniza o baseline de segurança em uma única linha.
 */
export class SecureBucket extends Construct {
  public readonly bucket: s3.Bucket;

  constructor(scope: Construct, id: string, props: SecureBucketProps) {
    super(scope, id);

    this.bucket = new s3.Bucket(this, 'Bucket', {
      bucketName: props.bucketName,
      encryption: s3.BucketEncryption.KMS,
      encryptionKey: props.encryptionKey,
      bucketKeyEnabled: true,
      versioned: true,
      enforceSSL: true,
      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
      objectOwnership: s3.ObjectOwnership.BUCKET_OWNER_ENFORCED,
      lifecycleRules: props.lifecycleRules,
      removalPolicy: props.retain ? cdk.RemovalPolicy.RETAIN : cdk.RemovalPolicy.DESTROY,
      autoDeleteObjects: !props.retain,
    });
  }
}

File: devops-agent-iac/abordagem-2-cdk-python/app_lib/constructs/secure_bucket.py
"""Construct L3 reutilizavel: baseline de seguranca de bucket S3."""
import aws_cdk as cdk
from aws_cdk import aws_kms as kms
from aws_cdk import aws_s3 as s3
from constructs import Construct


class SecureBucket(Construct):
    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        *,
        bucket_name: str,
        encryption_key: kms.IKey,
        retain: bool,
        lifecycle_rules: list[s3.LifecycleRule] | None = None,
    ) -> None:
        super().__init__(scope, construct_id)

        self.bucket = s3.Bucket(
            self,
            "Bucket",
            bucket_name=bucket_name,
            encryption=s3.BucketEncryption.KMS,
            encryption_key=encryption_key,
            bucket_key_enabled=True,
            versioned=True,
            enforce_ssl=True,
            block_public_access=s3.BlockPublicAccess.BLOCK_ALL,
            object_ownership=s3.ObjectOwnership.BUCKET_OWNER_ENFORCED,
            lifecycle_rules=lifecycle_rules,
            removal_policy=cdk.RemovalPolicy.RETAIN if retain else cdk.RemovalPolicy.DESTROY,
            auto_delete_objects=not retain,
        )
